extensions/notebook/src/book/bookTocManager.ts

```diff
@@ -190,54 +191,49 @@ export class BookTocManager implements IBookTocManager {
 	 * @param addSection The section that'll be added to the target section. If it's undefined then the target section (findSection) is removed from the table of contents.
 	*/
 	async updateTOC(version: BookVersion, tocPath: string, findSection: JupyterBookSection, addSection?: JupyterBookSection): Promise<void> {
-		const toc = yaml.safeLoad((await fs.readFile(tocPath, 'utf8')));
-		this._tocFiles.set(tocPath, toc);
-		let newToc = new Array<JupyterBookSection>(toc.length);
-		for (const [index, section] of toc.entries()) {
-			let newSection = this.buildTOC(version, section, findSection, addSection);
-			if (newSection) {
-				newToc[index] = newSection;
-			}
+		const tocFile = await fs.readFile(tocPath, 'utf8');
+		this.tableofContents = yaml.safeLoad(tocFile);
+		if (!this._tocFiles.has(tocPath)) {
+			this._tocFiles.set(tocPath, tocFile);
+		}
+		const isModified = this.modifyToc(version, this.tableofContents, findSection, addSection);
+		if (isModified) {
+			await fs.writeFile(tocPath, yaml.safeDump(this.tableofContents, { lineWidth: Infinity, noRefs: true, skipInvalid: true }));
+		} else {
+			throw (new Error(loc.sectionNotFound(findSection.title, tocPath)));
```

Comment:
What situations would this happen in? And is this going to provide enough information for the user to do something about it?

Reply:
This is an edge case, where the target section url is changed in the toc and the book tree view does not pick up the changes. So, if that happens then when trying to move a section/notebook then it'll modify the source toc file and remove the moving section, but it won't add it to the target section because it won't find it in the toc file. 
So, throwing an error would restore both toc files and then reinitialize the contents of the books.